datumaro/plugins/open_images_format.py

```diff
@@ -207,10 +217,144 @@ class OpenImagesImporter(Importer):
+        # the same thing.
+        root_node = {
+            # Create an OID-like label name that isn't already used by a real label
+            'LabelName': next(root_name
+                for i in itertools.count()
+                for root_name in [f'/m/{i}']
+                if root_name not in all_label_names
+            ),
+            # If an orphan has no children, then it makes no semantic difference
+            # whether it's listed in the hierarchy file or not. So strip such nodes
+            # to avoid recording meaningless data.
+            'Subcategory': [node for node in orphan_nodes if 'Subcategory' in node],
+        }
+
+        hierarchy_path = osp.join(
+            self._save_dir, OpenImagesPath.ANNOTATIONS_DIR, OpenImagesPath.HIERARCHY_FILE_NAME)
+
+        with open(hierarchy_path, 'w', encoding='utf-8') as hierarchy_file:
+            json.dump(root_node, hierarchy_file, indent=4, ensure_ascii=False)
+            hierarchy_file.write('\n')
+
+    def _save_subsets(self):
+        label_categories = self._extractor.categories().get(
+            AnnotationType.label, LabelCategories())
+
+        for subset_name, subset in self._extractor.subsets().items():
+            if _RE_INVALID_SUBSET.fullmatch(subset_name):
+                raise UnsupportedSubsetNameError(item_id=next(iter(subset)).id, subset=subset)
+
+            image_description_name = f'{subset_name}-images-with-rotation.csv'
+            image_description_fields = [
+                'ImageID',
+                'Subset',
+                'OriginalURL',
+                'OriginalLandingURL',
+                'License',
+                'AuthorProfileURL',
+                'Author',
+                'Title',
+                'OriginalSize',
+                'OriginalMD5',
+                'Thumbnail300KURL',
+                'Rotation',
+            ]
+
+            label_description_name = f'{subset_name}-annotations-human-imagelabels.csv'
+            label_description_fields =  [
+                'ImageID',
+                'Source',
+                'LabelName',
+                'Confidence',
+            ]
+
+            with \
+                self._open_csv_annotation(
+                    image_description_name, image_description_fields) as image_description_writer, \
+                self._open_csv_annotation(
+                    label_description_name, label_description_fields) as label_description_writer \
+            :
+                image_description_writer.writeheader()
+                label_description_writer.writeheader()
+
+                for item in subset:
+                    image_description_writer.writerow({
+                        'ImageID': item.id, 'Subset': subset_name,
+                    })
+
+                    if self._save_images and item.has_image:
+                        self._save_image(item, subdir=osp.join(
+                            OpenImagesPath.IMAGES_DIR, subset_name))
```

Comment:
```suggestion
                    if self._save_images:
                        if item.has_image and item.image.has_data:
                            self._save_image(item, subdir=osp.join(
                                OpenImagesPath.IMAGES_DIR, subset_name))
                        else:
                            log.debug("Item '%s' has no image", item.id)
```

Reply:
Doesn't `_save_image` produce this message by itself?

Reply:
Yes, it does. But if `item.has_image = False` then `_save_images` will not be called. Isn't it so?

Reply:
OK, I see what's going on. `_save_image` checks `item.image.has_data`, but it doesn't check `item.has_image`. So I can't just rely on it to output the warning.

This honestly seems like a design oversight to me, but for now I'll just implement your suggestion.

Reply:
Implemented.